extension/logstash/lib/logstash/util/retry_timer_task.rb

```diff
@@ -0,0 +1,31 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+require 'java'
+
+class RetryTimerTask < java.util.TimerTask
+   def initialize(retry_queue, event)
+      @retry_queue = retry_queue
+      # event style: [documents, http_headers, event_num, req_count]
+      @event = event
+      super()
+   end
+
+   def run
+      @retry_queue << @event
```

Comment:
1. Is there any limit on length of retry_queue?
2. Why not just add event to retry_queue but use a task and timer to do it?

Reply:
1. No. Because there won't be too many failures.
2. Because we need to use a timer to make the event wait for a certain amount of time before being queued and retried, rather than retrying immediately.